File: clients/python/shutdown.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on 2018-12-28
@author: Tobias Tangemann

Script for shuting down windows pc
"""
import socket
import hashlib
import hmac


def authenticated_response(challenge, key, force):
    """Calculate a response using the scret and the given challenge"""
    if len(challenge) < 32:
        raise Exception("Invalid challenge")
    if force:
        cmd = 'admin_shutdown'
    else:
        cmd = 'shutdown'
    message = f'{cmd}.{challenge}'.encode('ascii')
    mac = hmac.new(key.encode('ascii'), message, hashlib.sha256)
    return f'{cmd}.{challenge}.{mac.hexdigest()}'

def shutdown(host, port, secret, force, socket_timeout):
    """Send shutdown command to the specified host"""
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as conn:
        conn.settimeout(socket_timeout)
        try:
            # connect, request the challenge and send the signed response
            conn.connect((host, port))
            conn.send(b'request_challange\n')
            challenge = conn.recv(2048).decode('ascii').strip()
            response = authenticated_response(challenge, secret, force)
            conn.send((response + '\n').encode('ascii'))
            result = conn.recv(2048).decode('ascii').strip()
            if result == '1':
                return True, None
            return False, result
        except Exception as exp:
            return False, str(exp)
        finally:
            conn.close()

def main():
    """Main function"""
    result, error = shutdown('localhost', 10102, 'test', False, 30)

    if result:
        print("Shutdown successful")
    else:
        print(f'Error: {error}')


if __name__ == '__main__':
    main()

File: clients/go/shutdown.go
package main

import (
	"bufio"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"flag"
	"fmt"
	"net"
	"os"
	"strings"
	"time"
)

type params struct {
	host    string
	port    int
	token   string
	timeout int
	force   bool
	verbose bool
}

var config params

func main() {
	flag.StringVar(&config.token, "token", "", "Shared secret")
	flag.StringVar(&config.host, "host", "", "Adress/IP of windows pc")
	flag.IntVar(&config.port, "port", 10102, "Port")
	flag.IntVar(&config.timeout, "timeout", 3, "Timeout for request in seconds")
	flag.BoolVar(&config.verbose, "verbose", false, "Verbose mode")
	flag.Parse()

	// Create socket
	d := net.Dialer{Timeout: time.Duration(config.timeout) * time.Second}
	conn, err := d.Dial("tcp", fmt.Sprintf("%s:%d", config.host, config.port))
	checkError(err)

	// Request a challange
	_, err = conn.Write([]byte("request_challange\n"))
	checkError(err)

	reader := bufio.NewReader(conn)
	challange, err := reader.ReadString('\n')
	checkError(err)
	challange = strings.Trim(challange, "\n")

	var command string
	if config.force {
		command = "admin_shutdown."
	} else {
		command = "shutdown."
	}

	// Send the response
	_, err = conn.Write([]byte(command + authenticatedResponse(command+challange, config.token) + "\n"))
	checkError(err)

	// Get result of command
	result, err := reader.ReadString('\n')
	checkError(err)

	result = strings.Trim(result, "\n")
	fmt.Println(result)

	if result == "1" {
		os.Exit(0)
	} else {
		os.Exit(1)
	}
}

func authenticatedResponse(message string, key string) string {
	mac := hmac.New(sha256.New, []byte(key))
	mac.Write([]byte(message))
	return hex.EncodeToString(mac.Sum(nil))
}

func checkError(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "Fatal error: %s", err.Error())
		os.Exit(1)
	}
}
